third_party/amd/lib/TritonAMDGPUToLLVM/ConvertLayoutOpToLLVM.cpp

```diff
@@ -115,6 +115,102 @@ struct LocalLoadOpConversion
   }
 };
 
+struct ConvertLayoutOpMFMAToDotOpConversion
+    : public ConvertOpToLLVMPattern<triton::gpu::ConvertLayoutOp> {
+public:
+  explicit ConvertLayoutOpMFMAToDotOpConversion(
+      LLVMTypeConverter &typeConverter, const TargetInfoBase &targetInfo,
+      PatternBenefit benefit)
+      : ConvertOpToLLVMPattern<triton::gpu::ConvertLayoutOp>(typeConverter,
+                                                             benefit),
+        targetInfo(targetInfo) {}
+
+  LogicalResult
+  matchAndRewrite(triton::gpu::ConvertLayoutOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto srcType = cast<RankedTensorType>(op.getSrc().getType());
+    auto dstType = cast<RankedTensorType>(op.getType());
+
+    if (!matchMFMAAndDotOperandShuffleCase(srcType, dstType))
+      return failure();
+
+    /*
+    Using wave shuffle to convert layouts:
+    1) Input MMA layout (32x32, fp8, 16 values):
+     _____________________________________________________________
+    |(t0  v0 v1 v2 v3) (t32 v0 v1 v2 v3) ... (t32 v12 v13 v14 v15)|
+    | ...                                ...                      |
+    |(t31 v0 v1 v2 v3) (t63 v0 v1 v2 v3) ... (t63 v12 v13 v14 v15)|
+    |_____________________________________________________________|
+
+    2) Output Dot operand layout (two 32x16 tiles, fp8, 8 values each):
+     ____________________________________________________________   ___
+    |(t0  v0 v1 v2 v3 v4 v5 v6 v7) (t32 v0 v1 v2 v3 v4 v5 v6 v7) | |
+    | ...                           ...                          | |...
+    |(t31 v0 v1 v2 v3 v4 v5 v6 v7) (t63 v0 v1 v2 v3 v4 v5 v6 v7) | |
+    |____________________________________________________________| |___
+    */
+
+    auto loc = op.getLoc();
+
+    SmallVector<Value> inVals =
+        unpackLLElements(loc, adaptor.getSrc(), rewriter);
+    if (inVals.empty() || inVals.size() % 8 != 0)
+      return failure();
+
+    Value threadId = tid_val();
+    Value warpSize = i32_val(64); // MFMA Warp Size
+    Value laneId = urem(threadId, warpSize);
+    Value laneOffset = i32_val(32);
+    Value mask = icmp_slt(laneId, laneOffset);
+    Value addr0 = select(mask, add(laneId, laneOffset), laneId);
+    Value addr1 = select(mask, laneId, sub(laneId, laneOffset));
+
+    SmallVector<Value> outVals;
+    auto elemTy = int_ty(8);
+    auto vecTy = vec_ty(elemTy, 4);
+    for (size_t startIdx = 0; startIdx < inVals.size(); startIdx += 8) {
+      Value vec0 = undef(vecTy);
+      for (size_t vIdx = 0; vIdx < 4; ++vIdx) {
```

Comment:
Are all these hardcoded numbers, 32, 4, 8 here specific to mfma_32 instructions? Do you have a plan to support mfma_16 instructions?

Reply:
yeah, good point - right now, as I understand, we use mfma_16 only with very small shapes, defaulting to mfma_32 in general, so I left it as a follow up

Reply:
Actually, we use mfma_16 instruction for large compute bound gemms on MI300X. For FA, we use mfma_32 from the beginning before we have support for mfma_16. But later, we don't have time to test which one is better for FA.

Reply:
will add mfma_16 support